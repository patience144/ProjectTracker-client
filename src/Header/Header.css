header {
  display: flex;
  justify-content: space-between;
  align-items: center;
  width: calc(100% - 5% - 25px);
  height: 45px;
  max-height: 125px;
  padding: 0 40px 0 10px;
  max-height: 125px;
  z-index: 1;
  color: black;
  background-color: #c8529b;
  border-bottom: solid;
  border-width: 0.25px;
  border-color: rgba(255, 255, 255, 0.479);
}

h1 {
  color: black;
  text-decoration: none;
  font-size: 40px;
  font-family: "Cormorant Garamond", serif;
  margin-bottom: 25px;
  text-decoration: underline;
  text-decoration-color: rgb(95, 0, 95);
}

h4 {
  color: black;
  margin: 0;
  margin-top: 25px;
  font-size: 1.5vh;
}

.header {
  display: flex;
  justify-content: center;
  align-items: center;
  background-color: rgb(104, 104, 104);
  height: 75px;
  color: black;
  width: 100%;
  font-size: 2vh;
}

header > h1 {
  margin: 0;
  font-size: 3vh;
}

header ul {
  font-weight: bolder;
  display: flex;
}

header i {
  transform: scale(1.5);
}

nav ul li {
  font-size: 2.5vh;
  list-style-type: none;
  width: 135px;
}

nav ul li a {
  text-decoration: none;
  color: black;
}

nav ul {
  display: none;
}

nav {
  display: flex;
}

.nav {
  display: none;
  position: absolute;
  right: 0;
  height: 100%;
  width: 40%;
  background-color: #c8529b;
  z-index: 2;
}

.nav ul {
  display: flex;
  flex-direction: column;
  margin-top: 30%;
  padding-left: 10%;
}

.show-menu {
  display: block;
}

.faded::after {
  content: "";
  display: block;
  position: fixed;
  width: 100%;
  height: 100%;
  background: rgba(88, 88, 88, 0.616);
  top: 35px;
  top: 0;
  left: 0;
}

@media all and (min-width: 700px) {
  .header {
    height: 125px;
    justify-content: flex-start;
    padding-left: 5%;
    font-size: 2.5vh;
  }

  header {
    display: flex;
    align-items: flex-start;
    height: 75px;
    padding-right: 5%;
    padding: 5px 50px 10px 10px;
  }

  .container {
    margin-top: 50px;
    margin-bottom: 15px;
  }

  li a {
    margin-left: 35px;
  }

  h1 {
    font-size: 4.5vh;
  }

  header i {
    display: none !important;
  }

  nav ul {
    display: flex;
    flex-direction: row;
    align-items: center;
    height: 75px;
  }

  nav li {
    margin: 0;
  }
}

/* @media all and (min-width: 800px) {
  nav ul li {
    font-size: 3vh;
  }
} */
