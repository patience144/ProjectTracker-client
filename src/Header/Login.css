.login {
  display: flex;
  flex-direction: column;
  align-items: center;
  color: black;
  width: 50%;
}

.login input {
  height: 2.5em;
  border-radius: 5px;
  margin-top: 5px;
  width: 250px;
}

.login label {
  margin-top: 25px;
}

.login-buttons {
  display: flex;
  flex-direction: row;
  justify-content: space-around;
  width: 50%;
  height: 2em;
}

.login-buttons button {
  background-color: rgba(163, 22, 219, 0.925);
  border: none;
  width: 120px;
  height: 1.8em;
  outline: none;
  font-family: "Cormorant Garamond", serif;
  font-weight: bolder;
  font-size: 20px;
  color: rgb(95, 0, 95);
  border-radius: 3px;
}

.login-buttons button + button {
  margin-left: 20px;
}

.login button:hover {
  background-color: rgba(238, 56, 245, 0.699);
}

.login button:active {
  transform: scale(0.92);
}

.login h3 {
  margin-bottom: 50px;
}

span {
  margin-top: 10px;
  color: red;
}

@media all and (max-width: 400px) {
  .login {
    width: 100%;
  }

  .login h3 {
    font-size: 4vh;
    margin-bottom: 30px;
    border-color: black;
  }

  .login-buttons {
    align-items: center;
    justify-content: center;
    width: 100%;
    margin-top: 50px;
  }

  .login button {
    margin: 0 10px;
    min-height: 1.5em;
    width: 100px;
    height: 1.8em;
    font-size: 18px;
  }
}
