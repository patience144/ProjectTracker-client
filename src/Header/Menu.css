.header-options {
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  align-self: center;
  align-items: center;
  width: 50%;
  height: 50px;
  margin-bottom: 10px;
}

.header-options a,
.header-options a {
  color: black;
  text-decoration: none;
  font-size: 25px;
  font-weight: bold;
  font-family: "Cormorant Garamond", serif;
}

.header-options a {
  background-color: transparent;
  border: none;
  cursor: pointer;
}

.buttons {
  display: flex;
  flex-direction: row;
  justify-content: space-around;
  align-items: center;
  width: 110%;
  margin: 0;
}

.header-options h4 {
  font-size: 2.5vh;
}

.loggedIn {
  display: flex;
  flex-direction: row;
  align-items: center;
  height: 100px;
}

.loggedIn button {
  margin-bottom: 50px;
  margin-left: 25px;
  border: solid;
  color: black;
}

#users-button,
#account-button {
  margin-right: 15px;
}

.nav li + li {
  margin-top: 25%;
}

header h1 {
  font-size: 4vh;
}

header h4 {
  font-size: 2vh;
}

@media all and (max-width: 400px) {
  .header-options {
    align-self: flex-end;
    box-sizing: border-box;
    height: 50px;
    width: 65%;
  }

  .header-options a,
  .header-options a {
    font-size: 15px;
    margin: 0;
  }

  .buttons {
    align-items: flex-end;
  }

  .header-options h4 {
    font-size: 5.8vh;
  }
}
