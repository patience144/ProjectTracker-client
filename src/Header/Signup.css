.signup {
  display: flex;
  flex-direction: column;
  align-items: center;
  color: black;
  width: 50%;
  margin-top: 4%;
}

.signup h3 {
  text-align: center;
  font-size: 4vh;
  margin-bottom: 25px;
}

.signup input {
  height: 2em;
  margin-top: 3px;
  border-radius: 3px;
}

.signup label {
  margin-top: 25px;
}

.signup span {
  margin-top: 8px;
}

.signup-buttons {
  display: flex;
  flex-direction: row;
  justify-content: space-around;
  max-width: 350px;
  height: 2em;
  margin: 50px 10px;
  width: 100%;
}

.signup-buttons button {
  background-color: rgba(163, 22, 219, 0.925);
  border: none;
  width: 120px;
  height: 1.8em;
  outline: none;
  font-family: "Cormorant Garamond", serif;
  font-weight: bolder;
  font-size: 20px;
  color: rgb(95, 0, 95);
  border-radius: 3px;
}

.signup-buttons button + button {
  margin-left: 25px;
}

.signup button:hover {
  background-color: rgba(238, 56, 245, 0.699);
  border-color: rgba(238, 56, 245, 0.699);
}

.signup button:active {
  transform: scale(0.92);
}

#usernameTaken,
#emailTaken {
  color: red;
}

@media all and (max-width: 400px) {
  .signup {
    width: 100%;
  }

  .signup h3 {
    font-size: 3.2vh;
    margin-bottom: 30px;
  }

  .signup-buttons {
    align-items: center;
    justify-content: center;
    width: 100%;
    margin-top: 35px;
  }

  .signup button {
    margin: 50px 10px;
    min-height: 1.5em;
    width: 100px;
    height: 1.8em;
    font-size: 18px;
  }
}
