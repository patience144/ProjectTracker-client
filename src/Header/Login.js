import React from 'react';
import { withRouter } from 'react-router-dom';
import PropTypes from 'prop-types';
import api from '../api';
import './Login.css';

class Login extends React.Component {
  static propTypes = {
    setIdleTimer: PropTypes.func,
    updateUser: PropTypes.func
  };

  login = (evt) => {
    evt.preventDefault();
    const values = {
      username: evt.target.username.value,
      password: evt.target.password.value
    };

    for (const [key, value] of Object.entries(values)) {
      if (value === '') {
        evt.target[key].focus();
        document.getElementById('missing-values').style.display = 'inline-block';
        return;
      }
    };
    
    api.login(values)
      .then(res => {
        const authToken = res.apiKey || res.authToken;
        if (authToken) {
          /* If 'authToken' received and successful, set active User
            for app and begin refreshToken() timer. */
          window.sessionStorage.setItem('authToken', authToken);
          this.props.setIdleTimer();
          this.props.updateUser(res.user);
          this.props.history.push('/projects');
        } else {
          /* If no 'authToken', highlighth error label and 
            focus field to fix */
          const field = res.error.split(': ')[1];
          document.querySelectorAll('span').forEach(span => {
            span.style.display = 'none';
          });
          document.getElementById(field).focus();
          document.getElementById(`${field}Invalid`).style.display = 'inline-block';
        };
      });
  }

  render() {
    return (
      <form className='login' onSubmit={(evt) => this.login(evt)}>
        <h3>Login:</h3>
        <span style={{display: 'none'}} id='missing-values'>Missing values!</span>
        <label htmlFor='username'>Username:</label>
        <input type='text' name='username' id='username' autoComplete='username'/>
        <span style={{display: 'none'}} id='usernameInvalid'>Invalid username!</span>
        <label htmlFor='password'>Password:</label>
        <input type='password' name='password' id='password' autoComplete='current-password'/>
        <span style={{display: 'none'}} id='passwordInvalid'>Invalid password!</span>
        <div className='login-buttons'>
          <button type='submit'>Login</button>
          <button
            type='button'
            onClick={() => this.props.history.push('/')}
          >Cancel</button>
        </div>
      </form>
    );
  };

  componentDidMount() {
    const token = window.sessionStorage.getItem('authToken');
    if (token) this.props.history.push('/'); // Turn to HomePage if already logged in.
  };
};

export default withRouter(Login);