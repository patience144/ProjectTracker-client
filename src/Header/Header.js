import React from 'react';
import { Link } from 'react-router-dom';
import Menu from './Menu';
import './Header.css';

class Header extends React.Component {
  render() {
    const token = window.sessionStorage.getItem('authToken');
    const link = (token) ? '/projects' : '/';
    return (
      <header>
        <Link to={link} className='logo'><h1>ProjectTracker</h1></Link>
        <Menu />
      </header>
    );
  };
};

export default Header;