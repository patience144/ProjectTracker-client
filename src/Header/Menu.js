import React from "react";
import { Link, withRouter } from "react-router-dom";
import { API_KEY } from "../config";
import jwt_decode from "jwt-decode";
import UserContext from "../context/UserContext";
import "./Menu.css";

class Menu extends React.Component {
  state = { menu: false };

  static contextType = UserContext;

  render() {
    const menu = this.state.menu;
    const token = window.sessionStorage.getItem("authToken");
    const id = this.context ? this.context.id : null;
    let username = token ? this.context.username : null;
    const admin = username === "admin";
    if (!username && token && token !== API_KEY)
      username = jwt_decode(token).sub;
    return (
      <>
      {(username) ? <h4>Welcome, {username}.</h4> : null}
      <nav className={menu ? "faded" : ""}>
        <ul>
          <li><Link to="/signup">Sign-Up</Link></li>
          <li><Link to="/login">Login</Link></li>
          <li>
            <a
              href="https://github.com/patience144/projecttracker-client"
              target="_blank"
              rel="noreferrer"
            >
              GitHub
            </a>
          </li>
        </ul>
        <i
          onClick={(evt) => {
            evt.stopPropagation();
            this.setState({ menu: !menu });
          }}
          onBlur={this.blur}
          className="fa fa-bars"
        ></i>
        <div className={`nav ${menu ? "show-menu" : ""}`}>
          <ul>
            {
              // Show different access buttons depending if Logged in or not.
              !token ? (
                <>
                  <li>
                    <Link to="/signup">Sign-Up</Link>
                  </li>
                  <li>
                    <Link to="/login">Login</Link>
                  </li>
                </>
              ) : (
                admin
                  ? <li><Link id="users-button" to={`/users`}>Users</Link></li>
                  : <li><Link id="account-button" to={`/users/${id}`}>Accounts</Link></li>
              )
            }
            <li>
              <a
                href="https://github.com/patience144/projecttracker"
                target="_blank"
                rel="noreferrer"
              >
                GitHub
              </a>
            </li>
            {
              (!token)
                ? null
                : <li>
                    <Link
                      to="/logout"
                      id="logout-button"
                      onClick={() => {
                        window.sessionStorage.removeItem("authToken");
                        this.props.history.push("/");
                      }}
                    >
                      Log Out
                    </Link>
                  </li>
            }
          </ul>
        </div>
      </nav>
      </>
    );
  }

  componentDidMount() {
    document.addEventListener("click", () => this.setState({ menu: false }));
  }
}

export default withRouter(Menu);
