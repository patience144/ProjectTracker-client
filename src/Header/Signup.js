import React from 'react';
import { withRouter } from 'react-router-dom';
import PropTypes from 'prop-types';
import api from '../api';
import './Signup.css';

class Signup extends React.Component {
  static propTypes = { setIdleTimer: PropTypes.func };

  signup = (evt) => {
    evt.preventDefault();
    const user = {
      username: evt.target.username.value, firstName: evt.target.firstName.value,
      lastName: evt.target.lastName.value, email: evt.target.email.value, 
      tools: evt.target.tools.value, startDate: new Date(),
      github: evt.target.github.value, password: evt.target.password.value
    };
    
    for (const [key, value] of Object.entries(user)) {
      if (key === 'tools' || key === 'github' || key === 'start_date') continue;
      if (value === '') {
        evt.target[key].focus();
        document.getElementById('missing-values').style.display = 'inline-block';
        return;
      }
    };

    // Catch repeatPassword value without adding to user
    const repeatPassword = evt.target.repeatPassword.value;

    // Password validation
    if (user.password !== repeatPassword) {
      document.querySelectorAll('span').forEach(span => {
        span.style.display = 'none';
      });
      document.getElementById('password').focus();
      document.getElementById(`passwordError`).style.display = 'inline-block';
    } else {
      api.addUser(user)
        .then(res => {
          // Catch key and determine whether Admin or regular User
          const authToken = res.apiKey || res.authToken;
          if (authToken) {
            window.sessionStorage.setItem('authToken', authToken);
            this.props.setIdleTimer();
            this.props.updateUser(res.user);
            this.props.history.push('/projects');
          } else {
            // Catching field validation error from API error response
            const field = 
              res.error.includes('username')
                ? 'username'
                : res.error.includes('password')
                  ? 'password'
                  : 'email';
            document.querySelectorAll('span').forEach(span => {
              span.style.display = 'none';
            });
            // Show Error label and focus field with error.
            document.getElementById(field).focus();
            document.getElementById(`${field}Error`).style.display = 'inline-block';
          };
        })
        .catch(error => console.log({ error }));
    };
  };

  render() {
    const token = window.sessionStorage.getItem('authToken');
    if (token) this.props.history.goBack(); // Turn back if already logged in.
    return (
      <form className='signup' onSubmit={this.signup}>
        <h3>Sign-up for an account:</h3>
        <span style={{display: 'none'}} id='missing-values'>Missing values!</span>
        <label htmlFor='username'>Username:</label>
        <input type='text' name='username' id='username'  autoComplete='username'/>
        <span style={{display: 'none'}} id='usernameError'>Username taken!</span>
        <label htmlFor='firstName'>First Name:</label>
        <input type='text' name='firstName' id='firstName' />
        <label htmlFor='lastName'>Last Name:</label>
        <input type='text' name='lastName' id='lastName' />
        <label htmlFor='email'>E-mail:</label>
        <input type='text' name='email' id='email'/>
        <span style={{display: 'none'}} id='emailError'>E-mail taken!</span>
        <label htmlFor='password'>Password:</label>
        <input type='password' name='password' id='password' autoComplete='new-password'/>
        <label htmlFor='repeatPassword'>Repeat password:</label>
        <input type='password' name='repeatPassword' id='repeatPassword' autoComplete='new-password'/>
        <span style={{display: 'none'}} id='passwordError'>Passwords don't match!</span>
        <label htmlFor='tools'>Tools:</label>
        <input type='text' name='tools' id='tools' />
        <label htmlFor='github'>GitHub:</label>
        <input type='text' name='github' id='github' />
        <div className='signup-buttons'>
          <button type='submit'>Sign-Up</button>
          <button
            type='button'
            onClick={() => this.props.history.push('/')}
          >
            Cancel
          </button>
        </div>
      </form>
    );
  };
};

export default withRouter(Signup);