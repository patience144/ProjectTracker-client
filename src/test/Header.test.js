import { BrowserRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import ReactDOM from 'react-dom';
import Header from '../Header/Header';
import Signup from '../Header/Signup';
import Login from '../Header/Login';

describe('Header Section', () => {
  const header = <BrowserRouter><Header /></BrowserRouter>;
  const login = <BrowserRouter><Login /></BrowserRouter>;
  const signup = <BrowserRouter><Signup /></BrowserRouter>;
  describe('The Header', () => {
    it('renders without crashing', () => {
      const div = document.createElement('div');
      ReactDOM.render(header, div);
      ReactDOM.unmountComponentAtNode(div);
    });
    it('renders the UI as expected', () => {
      const tree = renderer.create(header).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
  describe('The Login page', () => {
    it('renders without crashing', () => {
      const div = document.createElement('div');
      ReactDOM.render(login, div);
      ReactDOM.unmountComponentAtNode(div);
    });
    it('renders the UI as expected', () => {
      const tree = renderer.create(login).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
  describe('The Sign-Up page', () => {
    it('renders without crashing', () => {
      const div = document.createElement('div');
      ReactDOM.render(signup, div);
      ReactDOM.unmountComponentAtNode(div);
    });
    it('renders the UI as expected', () => {
      const tree = renderer.create(signup).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});