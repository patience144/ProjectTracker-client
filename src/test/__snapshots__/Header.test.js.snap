// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header Section The Header renders the UI as expected 1`] = `
<header>
  <a
    className="logo"
    href="/"
    onClick={[Function]}
  >
    <h1>
      #DevMap
    </h1>
  </a>
  <div
    className="header-options"
  >
    <div
      className="buttons"
    >
      <a
        href="/signup"
        onClick={[Function]}
      >
        Sign-Up
      </a>
      <a
        href="/login"
        onClick={[Function]}
      >
        Login
      </a>
    </div>
  </div>
</header>
`;

exports[`Header Section The Login page renders the UI as expected 1`] = `
<form
  className="login"
  onSubmit={[Function]}
>
  <h3>
    Login:
  </h3>
  <span
    id="missing-values"
    style={
      Object {
        "display": "none",
      }
    }
  >
    Missing values!
  </span>
  <label
    htmlFor="username"
  >
    Username:
  </label>
  <input
    autoComplete="username"
    id="username"
    name="username"
    type="text"
  />
  <span
    id="usernameInvalid"
    style={
      Object {
        "display": "none",
      }
    }
  >
    Invalid username!
  </span>
  <label
    htmlFor="password"
  >
    Password:
  </label>
  <input
    autoComplete="current-password"
    id="password"
    name="password"
    type="password"
  />
  <span
    id="passwordInvalid"
    style={
      Object {
        "display": "none",
      }
    }
  >
    Invalid password!
  </span>
  <div
    className="login-buttons"
  >
    <button
      type="submit"
    >
      Login
    </button>
    <button
      onClick={[Function]}
      type="button"
    >
      Cancel
    </button>
  </div>
</form>
`;

exports[`Header Section The Sign-Up page renders the UI as expected 1`] = `
<form
  className="signup"
  onSubmit={[Function]}
>
  <h3>
    Sign-up for an account:
  </h3>
  <span
    id="missing-values"
    style={
      Object {
        "display": "none",
      }
    }
  >
    Missing values!
  </span>
  <label
    htmlFor="username"
  >
    Username:
  </label>
  <input
    autoComplete="username"
    id="username"
    name="username"
    type="text"
  />
  <span
    id="usernameError"
    style={
      Object {
        "display": "none",
      }
    }
  >
    Username taken!
  </span>
  <label
    htmlFor="firstName"
  >
    First Name:
  </label>
  <input
    id="firstName"
    name="firstName"
    type="text"
  />
  <label
    htmlFor="lastName"
  >
    Last Name:
  </label>
  <input
    id="lastName"
    name="lastName"
    type="text"
  />
  <label
    htmlFor="email"
  >
    E-mail:
  </label>
  <input
    id="email"
    name="email"
    type="text"
  />
  <span
    id="emailError"
    style={
      Object {
        "display": "none",
      }
    }
  >
    E-mail taken!
  </span>
  <label
    htmlFor="password"
  >
    Password:
  </label>
  <input
    autoComplete="new-password"
    id="password"
    name="password"
    type="password"
  />
  <label
    htmlFor="repeatPassword"
  >
    Repeat password:
  </label>
  <input
    autoComplete="new-password"
    id="repeatPassword"
    name="repeatPassword"
    type="password"
  />
  <span
    id="passwordError"
    style={
      Object {
        "display": "none",
      }
    }
  >
    Passwords don't match!
  </span>
  <label
    htmlFor="tools"
  >
    Tools:
  </label>
  <input
    id="tools"
    name="tools"
    type="text"
  />
  <label
    htmlFor="github"
  >
    GitHub:
  </label>
  <input
    id="github"
    name="github"
    type="text"
  />
  <div
    className="signup-buttons"
  >
    <button
      type="submit"
    >
      Sign-Up
    </button>
    <button
      onClick={[Function]}
      type="button"
    >
      Cancel
    </button>
  </div>
</form>
`;
